Add noise based wind

cullvox opened the pull request:
This is a basic, working, implementation of wind. 

It's probably pretty inefficient at scale and might need optimizations.

- Wind can be enabled and disabled
- Noise scale and wind scale for both more control over simulation
- Runtime configuration of wind noise with FastNoiseLite settings

mphe reviewed (changes requested):
Thanks!
Besides these nitpicks, it looks fine.

The direction vector of the wind direction could be cached in GDScript so it doesn't need to be recomputed during simulation.
But I think it's negligible for now. 

I also think it would be a good idea to extract wind simulation parameters in a dedicated `Resource` to make it reusable and reduce memory usage with many ropes sharing the same parameters. I also considered setting it globally in `NativeRopeServer` to be applied to all ropes. But using Resources enables more flexibility.

cullvox commented on demo/addons/ropesim/Rope.gd:
```diff
@@ -41,6 +43,21 @@ signal on_point_count_changed()
+## Enables and disables wind simulation.
+@export var wind_enabled: bool = false
+
+## Scale that the strength can take
```
Yeah, so wind strength is the strength of the force almost like a gravitational force that just throws the rope in a particular direction, then the noise strength adds that realistic wave, and you can vary its intensity depending on how 'turbulent' you want the wind to look. I'll make these explain that better.

cullvox commented on src/NativeRopeContext.cpp:
```diff
@@ -144,6 +151,21 @@ void NativeRopeContext::_simulate_velocities(double delta)
+{
+    if (!wind_enabled || wind_noise.is_null()) return;
+
+    float time = (float)Time::get_singleton()->get_ticks_msec();
+    Vector2 dir = Vector2::from_angle(wind_direction);
```
Using a resource is a fantastic idea, I'm doing some research into how I can create resources for C++ and GDScript like this. I'll get working on that tonight. Using a resource makes wind signaling a little simpler too, that's what I was just dealing with in my project.

cullvox replied:
That was supposed to be a comment on the overall pull, oh well.

mphe replied:
There are two ways essentially. The first is creating the resource natively in C++ or doing it as with the rope node and define it in GDScript and sync the data over when needed.
Not sure which is better though in this case.
The GDScript approach always has the advantage of an easier @export interface and it allows simple modifications as long as they do not affect the backend.

mphe commented:
Btw, I also suggest using the clangd  language server as it has clang-tidy checks built-in which give you valuable static analysis hints and warnings.

cullvox commented:
Thanks for that suggestion, I am now setup with clangd. Usually I would use it, but I'm still getting familiar with the Godot scons build environment, so now everything should be setup as normal.

cullvox commented:
With the latest commit, I added a `NativeRopeWind` resource and updated the surrounding files for the change. I don't think that `NativeRopeWind` is a good name, we can come up with something better sounding. 

Eventually, a noise image should be generated if we want to improve performance. The wind resource solves a couple challenges regarding that.

mphe commented on src/NativeRopeContext.cpp:
```diff
@@ -144,6 +145,25 @@ void NativeRopeContext::_simulate_velocities(double delta)
+
+    for (int i = 0; i < points.size(); ++i)
+    {
+        const float noise = wind->noise->get_noise_3d(points[i].x, points[i].y, time);
```
That's a good idea! I first thought about projecting the point position vector on the wind direction but using  `get_noise_3d` is much simpler and straight forward.

mphe commented on src/NativeRopeContext.hpp:
```diff
@@ -1,8 +1,10 @@
 #include "godot_cpp/classes/curve.hpp"
 #include "godot_cpp/classes/node2d.hpp"
 #include "godot_cpp/classes/physics_shape_query_parameters2d.hpp"
+#include "godot_cpp/classes/fast_noise_lite.hpp"
```
Unused include

mphe commented on src/NativeRopeContext.hpp:
```diff
@@ -39,22 +42,24 @@ class NativeRopeContext  // NOLINT(cppcoreguidelines-special-member-functions)
     float max_endpoint_distance = 0.0;
     int num_constraint_iterations = 0;
-
+    
```
Trailing whitespace

mphe commented on src/NativeRopeContext.hpp:
```diff
@@ -39,22 +42,24 @@ class NativeRopeContext  // NOLINT(cppcoreguidelines-special-member-functions)
     float collision_damping = 0.0;
     int collision_mask = 0;
-
+    
```
Trailing whitespace

mphe commented on src/NativeRopeWind.cpp:
```diff
@@ -0,0 +1,68 @@
+#include "NativeRopeWind.hpp"
+#include "godot_cpp/classes/global_constants.hpp"
+#include "godot_cpp/core/class_db.hpp"
+#include "godot_cpp/core/math.hpp"
```
Unused include

mphe commented on src/NativeRopeWind.cpp:
```diff
@@ -0,0 +1,68 @@
+#include "godot_cpp/core/math.hpp"
+#include "godot_cpp/core/object.hpp"
+#include "godot_cpp/core/property_info.hpp"
+#include "godot_cpp/variant/utility_functions.hpp"
```
Unused include

mphe commented on src/NativeRopeWind.hpp:
```diff
@@ -0,0 +1,41 @@
+#define NATIVE_ROPE_WIND
+
+#include "godot_cpp/classes/fast_noise_lite.hpp"
+#include <godot_cpp/classes/resource.hpp>
```
Either stick to <> or "".

mphe commented on src/NativeRopeWind.hpp:
```diff
@@ -0,0 +1,41 @@
+        GDCLASS(NativeRopeWind, Resource) // NOLINT
+
+        public:
+
+            /// This is a member comment! 
```
Redundant newline and comment

mphe commented on src/NativeRopeWind.hpp:
```diff
@@ -0,0 +1,41 @@
+
+            /// This is a member comment! 
+            Vector2 direction = {1.0f, 0.0f};
+            float strength;
+            float noise_strength;
```
Missing default values

mphe commented on src/NativeRopeWind.hpp:
```diff
@@ -0,0 +1,41 @@
+            static void _bind_methods();
+
+        private:
+            float angle;
```
Missing default value

mphe commented on src/NativeRopeWind.hpp:
```diff
@@ -0,0 +1,41 @@
+            float noise_strength;
+            Ref<FastNoiseLite> noise;
+
+            NativeRopeWind() = default;
+            ~NativeRopeWind() override = default;
```
Can be removed since they are default by default

mphe commented on src/NativeRopeContext.cpp:
```diff
@@ -144,6 +145,25 @@ void NativeRopeContext::_simulate_velocities(double delta)
+        // Takes the perpendicular angle of the winds direction and multipilies it by the noise strength
+        const Vector2 noise_velocity = Vector2(wind->direction.y, -wind->direction.x) * (noise * wind->noise_strength);
+        
+        (*velocities)[i] += wind_velocity + noise_velocity;
```
I don't understand why there is a `noise_velocity` and `wind_velocity` and why they are added.
`wind_velocity` is just a constant offset regardless of the noise. If that is the intended purpose, rename `strength` to `bias` or `minimum_strength` or something that expresses the intent better.

Why is the actual noise applied in perpendicular wind direction?

cullvox replied:
The wind vector, pulls and makes the rope more taught towards the direction. The noise oscillates the point back and forth.

mphe replied:
I see. Ok, then let's rename `strength` -> `wind_strength` and `noise_strength` -> `oscillation_strength`.
Btw, you can use [`Vector2.orthogonal`](https://docs.godotengine.org/en/stable/classes/class_vector2.html#class-vector2-method-orthogonal).

You should also cache `wind_velocity` and `wind_dir.orthogonal() * oscillation_strength` right under `time` as these terms remain constant over the loop.
Missed that one before.

mphe commented on doc_classes/NativeRopeWind.xml:
```diff
@@ -0,0 +1,20 @@
+	<tutorials>
+	</tutorials>
+	<members>
+		<member name="direction" type="float" setter="set_direction" getter="get_direction" default="1.12104e-43">
```
Default value seems wrong

mphe commented on doc_classes/NativeRopeWind.xml:
```diff
@@ -0,0 +1,20 @@
+		</member>
+		<member name="noise_strength" type="float" setter="set_noise_strength" getter="get_noise_strength" default="0.0">
+		</member>
+		<member name="strength" type="float" setter="set_strength" getter="get_strength" default="1.61149e-43">
```
Default value seems wrong

mphe reviewed (changes requested):
In addition to the remarks:
- Use an underscore prefix for private and protected members and functions. Not for function arguments.
- Class declaration order should be: methods before attributes and public before protected before private (see other files)
- Resources should emit the `changed` signal when changed. See [emit_changed](https://docs.godotengine.org/en/stable/classes/class_resource.html#class-resource-method-emit-changed)
- Rename `NativeRopeWind` to `RopeWindParameters`, `RopeWindSettings`, `RopeWindProperties` or similar.
- Attributes of `NativeRopeWind` should be private because there are getters and setters.
- The wind example looks nice, but text needs a spell check. Also use ":" instead of "-" please. Description for strength is unclear.

cullvox commented:
Thanks for going through all this again! I'll try to make it a little cleaner next time.

cullvox commented on demo/addons/ropesim/Rope.gd:
```diff
@@ -41,6 +43,12 @@ signal on_point_count_changed()
 @export var gravity_direction: Vector2 = Vector2.DOWN
 
+## Enables and disables wind simulation.
+@export var wind_enabled: bool = false
```
Not really, it just makes it easier in GDScript. Just a quick `rope.wind_enabled = true` instead of having to have a reference to a wind object. Another reason is that was one for gravity, so I just thought I'd have another for wind. I'll remove it if you think it's redundant.

cullvox replied:
I have been thinking about this in the background all day. I think it's reasonable to say either one would work, I could see the instance where you might want to have it in the resource to enable/disable so you can easily turn it off for special events in a game. Otherwise, I can't think of a use case for individual turn off. Even then during individual turn off they could just remove the wind resource instead.

I believe I'm going to go for adding it to the resource, it's more useful.

cullvox commented:
I've still got to properly add documentation, but everything seems to be working as intended. I've still got to add the boolean enable value to the wind parameter resource, I'll add that tomorrow.

cullvox commented:
I've got this to a state that I like again. I added batch scripts for compiling on Windows too and they both work. Enabling and disabling wind is also in the resource now.

mphe commented on SConstruct:
```diff
@@ -26,7 +26,8 @@ opts.Update(env)
 # - LINKFLAGS are for linking flags
 
-env.Append(CCFLAGS="-fdiagnostics-color")
+if ["platform"] == "linux":
```
You need to check the host os, not the target platform. You can perfectly cross-compile on linux for windows with this compile flag.
See [`os.name`](https://docs.python.org/3/library/os.html#os.name).

mphe commented on demo/addons/ropesim/Rope.gd:
```diff
@@ -41,6 +43,12 @@ signal on_point_count_changed()
 @export var gravity_direction: Vector2 = Vector2.DOWN
 
+## Enables and disables wind simulation.
+@export var wind_enabled: bool = false
```
Unused now

mphe commented on demo/addons/ropesim/Rope.gd:
```diff
@@ -41,6 +43,12 @@ signal on_point_count_changed()
+## Enables and disables wind simulation.
+@export var wind_enabled: bool = false
+
+## Strength of wind force to keep it in a particular direction.
```
Update doc string to fit the property, e.g. "Optional parameters for wind simulation."

mphe commented on doc_classes/RopeWindParameters.xml:
```diff
@@ -0,0 +1,27 @@
+			Noise function used when generating oscillating wind forces.
+		</member>
+		<member name="oscillation_strength" type="float" setter="set_oscillation_strength" getter="get_oscillation_strength" default="0.0">
+			Amplitude of oscillations in the wind noise. Higher values increase the perpendicular distance from the wind direction.
```
perpendicular oscillations

cullvox replied:
I am assuming you want me to change from: <br>`"Amplitude of oscillations in the wind noise. Higher values increase the perpendicular distance from the wind direction."`<br> to  <br>`"Amplitude of oscillations in the wind noise. Higher values increase the perpendicular oscillations in the wind's direction."`

Is that right?

mphe commented on src/NativeRopeContext.cpp:
```diff
@@ -144,6 +143,21 @@ void NativeRopeContext::_simulate_velocities(double delta)
+    for (int i = 0; i < points.size(); ++i)
+    {
+        const float noise = wind->get_noise()->get_noise_3d(points[i].x, points[i].y, time);
+        const Vector2 noise_velocity = wind->get_direction_vector().orthogonal() * (noise * wind->get_oscillation_strength());
```
- `wind->get_direction_vector().orthogonal() * wind->get_oscillation_strength()` can be cached just like `wind_velocity`.
- Linebreak between the variables and the for loop

mphe commented on src/RopeWindParameters.cpp:
```diff
@@ -0,0 +1,99 @@
+
+void RopeWindParameters::set_direction(float angle)
+{
+    if (_angle == angle) 
```
Trailing whitespace

mphe commented on src/RopeWindParameters.cpp:
```diff
@@ -0,0 +1,99 @@
+    emit_changed();
+}
+
+bool RopeWindParameters::get_enable_wind()
```
Method can be const

mphe commented on src/RopeWindParameters.cpp:
```diff
@@ -0,0 +1,99 @@
+    return _noise;
+}
+
+void RopeWindParameters::set_enable_wind(bool enable_wind)
```
Use the same argument name as in the header for consistency.

mphe reviewed (changes requested):
Good work so far!

Regarding `godot-cpp` and documentation support. Rereading the docs it seems that at least `godot-cpp` 4.3 is needed to compile with documentation support, but I'm unsure if it would still be compatible with lower Godot versions. The documentation only talks about that if clause in the build script which makes it compatible with older build pipelines.
  - Let's just make it simple and bump the minimum version to 4.3. Ignore what I said before, sorry.
  - Track the `4.3` branch in `godot-cpp`

Thanks for unifying my inconsistent ""/<> includes as well :D 

I think that will be the last iteration until it's ready to merge.

cullvox committed:
Add noise based wind

mphe commented:
Thanks!
